Add FileDescriptor trait to abstract fn's on File's and Stdin,Stdout,Stderr

samrat opened the pull request:
Related issue: #1486 

Instead of mapping FDs to `FileHandle`, map them to a `FileDescriptor` trait object. The goal is to eventually have both `FileHandle` as well as `Stdin`, `Stdout` and `Stderr` implement this trait so that syscalls involving FDs can handle both `File`s as well as the standard IO streams.

This PR adds the `FileDescriptor` trait and an `impl` for `FileHandle`. I'll open a separate PR for implementing the trait for the standard IO streams.

samrat commented:
This is what the `impl` for `Stdin` would look like:

https://github.com/samrat/miri/commit/2c868a1eaa88c39806a9f2f3a9178782f24c961c#diff-560909e2fdd6fac0fb0b832d9470ac7cR56

(Of course, this is contingent on feedback on this PR)

oli-obk commented on src/shims/posix/fs.rs:
```diff
@@ -22,15 +22,42 @@ struct FileHandle {
+    fn as_file_handle(&self) -> InterpResult<'tcx, &FileHandle>;
+
+    fn read(&mut self, bytes: &mut [u8]) -> Result<usize, io::Error>;
+    fn write(&mut self, bytes: &[u8]) -> Result<usize, io::Error>;
```
I think these should return `InterpResult`, so that stdin can just error when being written to. This way we could remove the final `if fd == 1 || fd == 2 { throw_...` code, too

samrat replied:
I could change this. What would the appropriate error value be?

Currently in https://github.com/rust-lang/miri/pull/1495/files#diff-560909e2fdd6fac0fb0b832d9470ac7cL481 , we set the error code and return `Ok(-1)`. However, with this PR we don't have a way of setting the error code from `FileDescriptor` trait object currently.

oli-obk replied:
Hmm... what about returning a `Result<Option<usize>, io::Error>` then?

oli-obk replied:
Oh no, that makes no sense either. I think we need `InterpResult<'tcx, std::io::Result<usize>>`. This way you can `?` away the undefined impls and still use the same error processing we had

oli-obk commented on src/shims/posix/fs.rs:
```diff
@@ -412,24 +447,29 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
-                // discussion.
-                drop(file);
-                Ok(0)
+        if let Some(file_descriptor) = this.machine.file_handler.handles.remove(&fd) {
```
Why does `as_file_handle` return a `Result`? If it returned an `Option, you could
```suggestion
        if let Some(file_descriptor) = this.machine.file_handler.handles.remove(&fd).and_then(FileDescriptor::as_file_handle) {
```

and save yourself the additional `this.handle_not_found()`

oli-obk replied:
resolving, because it just makes things more complex

oli-obk commented on src/shims/posix/fs.rs:
```diff
@@ -383,20 +410,28 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             };
             let fd_result = file_result.map(|duplicated| {
-                fh.insert_fd_with_min_fd(FileHandle { file: duplicated, writable }, start)
+                fh.insert_fd_with_min_fd(FileHandle { file: duplicated, writable: writable }, start)
```
I don't think this change is necessary

oli-obk commented on src/shims/posix/fs.rs:
```diff
@@ -383,20 +410,28 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             let (file_result, writable) = match fh.handles.get(&fd) {
-                Some(FileHandle { file, writable }) => (file.try_clone(), *writable),
+                Some(file_descriptor) => match file_descriptor.as_file_handle() {
+                    Ok(FileHandle { file, writable }) => (file.try_clone(), writable.clone()),
```
why `.clone()` on `writable`, isn't that a `bool`?

oli-obk commented:
Do we need the `writeable` field anymore at all? Since we can now error out in the specific impl's `write` or `read` function, that should suffice, right?

samrat commented:
@oli-obk It is possible to open a `File` in read-only mode, so we have to error out on that too. 

If we remove the `writable`, I guess we could always just call `File::sync_all` and `File::set_len`(these seem to be the two methods guarded by `writable`) and let the OS tell us that the `File`s are readable?

oli-obk commented:
That sounds good, yes!

samrat commented:
I'm a bit confused about `maybe_sync_file` in https://github.com/rust-lang/miri/blob/master/src/shims/posix/fs.rs#L238

Looks like it always returns `0i32` regardless of what the `operation` fn returns. It should still be okay to attempt to sync for Windows as well, right?

oli-obk commented:
the operation is not invoked at all on windows, so we need to keep not invoking it.

samrat commented:
Hmm, in that case I don't think we can remove `writable`. It seems possible to get the same info from https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.permissions , but it might be simpler to just keep `writable`.

Did you have any other ideas on how we can remove `writable`?

oli-obk commented:
Hmm... I don't have any ideas right now that aren't worse than the flag ^^

Let's keep it as is for now

oli-obk commented:
So, then it's just the two nits I commented on, after that we can merge this

oli-obk commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 79e066fc95c036e64716a4222c580782a9c932c2 has been approved by `oli-obk`

<!-- @bors r=oli-obk 79e066fc95c036e64716a4222c580782a9c932c2 -->
<!-- homu: {"type":"Approved","sha":"79e066fc95c036e64716a4222c580782a9c932c2","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 79e066fc95c036e64716a4222c580782a9c932c2 with merge 69d050fb6b2a6d8ec0c01a44ccfaf79409c4b446...
<!-- homu: {"type":"BuildStarted","head_sha":"79e066fc95c036e64716a4222c580782a9c932c2","merge_sha":"69d050fb6b2a6d8ec0c01a44ccfaf79409c4b446"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/github/rust-lang/miri/builds/178281967), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/miri/builds/34454226)
Approved by: oli-obk
Pushing 69d050fb6b2a6d8ec0c01a44ccfaf79409c4b446 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/miri/builds/34454226","checks-travis":"https://travis-ci.com/github/rust-lang/miri/builds/178281967"},"merge_sha":"69d050fb6b2a6d8ec0c01a44ccfaf79409c4b446"} -->

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -22,15 +22,42 @@ struct FileHandle {
     writable: bool,
 }
 
+trait FileDescriptor<'tcx> : std::fmt::Debug {
```
Why does the trait depend on `'tcx`? That should not be necessary. It would be better to make each method take a `'tcx` lifetime parameter.

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -383,7 +410,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
-                Some(FileHandle { file, writable }) => (file.try_clone(), *writable),
+                Some(file_descriptor) => match file_descriptor.as_file_handle() {
+                    Ok(FileHandle { file, writable }) => (file.try_clone(), *writable),
+                    Err(_) => return this.handle_not_found(),
```
This is fishy -- there are very few situations where it is a good idea to inspect an `InterpResult` with `match`.

But also, the error is wrong. The handle *was* found, it just did not support this operation. So IMO it would be reasonable to just do `file_descriptor.as_file_handle()?` here.

RalfJung replied:
Oh and please add a FIXME saying that in the future, we'd of course like to support "dup" for *all* FDs, not just the ones with an underlying `FileHandle`.

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -394,9 +424,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
-                let io_result = maybe_sync_file(file, *writable, File::sync_all);
-                this.try_unwrap_io_result(io_result)
+            if let Some(file_descriptor) = this.machine.file_handler.handles.get(&fd) {
+                match file_descriptor.as_file_handle() {
```
Same, this should use `?` to produce the right error ("interpreter does not support this operation", instead of an incorrect "handle not found").

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -412,24 +447,29 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
-                drop(file);
-                Ok(0)
+        if let Some(file_descriptor) = this.machine.file_handler.handles.remove(&fd) {
+            match file_descriptor.as_file_handle() {
```
Same here.

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -1103,21 +1148,26 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
-                    this.set_last_error(einval)?;
-                    Ok(-1)
+        if let Some(file_descriptor) = this.machine.file_handler.handles.get_mut(&fd) {
+            match file_descriptor.as_file_handle() {
```
And same here.

RalfJung commented on src/shims/posix/fs.rs:
```diff
@@ -1239,7 +1304,10 @@ impl FileMetadata {
         let option = ecx.machine.file_handler.handles.get(&fd);
         let file = match option {
-            Some(FileHandle { file, writable: _ }) => file,
+            Some(file_descriptor) => match file_descriptor.as_file_handle() {
```
And here.

RalfJung commented:
And yes "writable" is needed on Windows only, because there syncing somehow fails for read-only files...

samrat commented:
@RalfJung I'll open a follow-up PR to address your comments.

RalfJung commented:
@samrat awesome, thanks!

What is your timeline for these? I'd like to get at least the error catching fixed soon to make sure Miri is in a clean state there. If you cannot get around to do that this week-end, that is completely okay -- then I would go ahead and do that part myself.

samrat commented:
@RalfJung I've opened a PR addressing the comments on error-handling. I'll try to address the rest of the comments in the next couple of days. Does that work?

RalfJung commented:
The rest is not very urgent, so that's great, yes. :)
I might make a PR for the isolation check, that one also makes me a bit nervous.
